DIPs/DIP1xxx-rval_to_ref.md

```diff
@@ -0,0 +1,345 @@
+https://forum.dlang.org/thread/mailman.1478.1521842510.3374.digitalmars-d@puremagic.com  
+https://forum.dlang.org/thread/gsdkqnbljuwssslxuglf@forum.dlang.org  
+
+Issues:  
+https://issues.dlang.org/show_bug.cgi?id=9238  
+https://issues.dlang.org/show_bug.cgi?id=8845  
+https://issues.dlang.org/show_bug.cgi?id=6221  
+https://issues.dlang.org/show_bug.cgi?id=6442  
+
+PRs:  
+https://github.com/dlang/dmd/pull/4717  
+
+## Contents
+* [Rationale](#rationale)
+* [Proposal](#proposal)
+* [Temporary destruction]()
+* [`@safe`ty implications](#safety_implications)
+* [Why not `auto ref`?](#auto_ref)
+* [Key use cases](#use_cases)
+* [Reviews](#reviews)
+
+## Rationale
+
+When calling functions that receive `ref` args, D prohibits supplying rvalues. It is suggested that this is to assist the author identifying likely logic errors where an rvalue will expire at the end of the statement, and it doesn't make much sense for a function to mutate a temporary whose life will not extend beyond the function call in question.
+
+However, many functions receive arguments by reference, and this may be for a variety of reasons.  
+One common reason is that the cost of copying large structs via the parameter list is expensive, so struct parameters may be received by reference to mitigate this cost.  
+Another common case is that the function may want to mutate the caller's data directly or return data via `out` parameters due to ABI limitations regarding multiple return values. This is the potential error case that the existing design attempts to mitigate, but in D, pipeline programming is vary popular, and contrary to conventional wisdom where the statement is likely to end at the end of the function call, pipeline expressions may result in single statements performing a lot of work, mutating state as it passes down the pipeline.
+
+A related issue is with relation to generic code which reflects or received a function by alias. Such generic code may want to call that function, but it is often the case that details about the ref-ness of arguments lead to incorrect semantic expressions in the generic code depending on the arguments, necessitating additional compile-time logic to identify the ref-ness of function arguments and implement appropriate workarounds on these conditions. This leads to longer, more brittle, and less-maintainable generic code. It is also much harder to write correctly the first time, and such issues may only emerge in niche use cases at a later time.
+
+With these cases in mind, the existing rule feels out-dated or inappropriate, and the presence of the rule may often lead to aggravation while trying to write simple, readable code.  
+Calling a function should be simple and orthogonal, generic code should not have to concern itself with details about ref-ness of function parameters, and users should not be required to jump through hoops when `ref` appears in API's they encounter.
+
+Consider the example:
```

Comment:
You examples should be collocated above with the issues they are intended to demonstrate, e.g.

"The first issue is yadda yadda. 

Example Here

A related issue is yadda yadda.

Example here